source/common/http/filter_manager.cc

```diff
@@ -404,11 +409,22 @@ void ActiveStreamDecoderFilter::requestDataTooLarge() {
   }
 }
 
-void FilterManager::addStreamDecoderFilterWorker(
-    StreamDecoderFilterSharedPtr filter, Matcher::MatchTreeSharedPtr<HttpMatchingData> matcher,
-    HttpMatchingDataImplSharedPtr matching_data, bool dual_filter) {
-  ActiveStreamDecoderFilterPtr wrapper(new ActiveStreamDecoderFilter(
-      *this, filter, std::move(matcher), std::move(matching_data), dual_filter));
+void FilterManager::addStreamDecoderFilterWorker(StreamDecoderFilterSharedPtr filter,
+                                                 FilterMatchStateSharedPtr match_state,
+                                                 bool dual_filter) {
+  ActiveStreamDecoderFilterPtr wrapper(
+      new ActiveStreamDecoderFilter(*this, filter, match_state, dual_filter));
+
+  // If we're a dual handling filter, have the encoding wrapper be the only thing registering itself
+  // as the handling filter.
```

Comment:
I'm a little confused by this. Doesn't this need to be configurable or potentially context specific? Or the idea here is that a single filter registered twice would then only get the callbacks once and it doesn't actually matter whether they are in the encoding or decoding path? And this is really only needed for state that has to go through to the underlying impl like skip? Is skip the only case where this matters?

Given that the pointers are going to be the same if they are both set (right?), could the skip case be handled by just calling a method on the filter base which somehow knows how to set skip for both encoding and decoding direction?

Reply:
The problem is that we have two distinct filter wrappers around the same underlying filter: https://github.com/envoyproxy/envoy/blob/master/source/common/http/filter_manager.h#L705-L708

As a result, we need both wrappers to know about the fact that a skip action has been evaluated, since the skip action isn't associated with a specific filter wrapper, but to the filter itself. This here is one strategy where we have the FilterMatchState responsible for doing the matching and setting the skip_ flag on each of the wrappers. I can imagine another approach where we instead have the wrappers check the FilterMatchState to see whether it should skip? 

For the filter action in the dual filter case we have two wrappers but we only want to notify the filter once about the match action, so we can get away with just arbitrarily picking one of the wrappers to use. Alternatively we could just hold onto a reference to the underlying filter instead: that way we're bypassing the wrappers in this case and things might be a bit clearer?

Reply:
> I can imagine another approach where we instead have the wrappers check the FilterMatchState to see whether it should skip?

I think this is what I would do since the state is shared?

> Alternatively we could just hold onto a reference to the underlying filter instead: that way we're bypassing the wrappers in this case and things might be a bit clearer?

Yeah +1.

So yeah personally I would make the above changes but if you don't can you add some more comments somewhere leaving a trail about this?

/wait